### ISSUE TITLE
neocomplete#enable_insert_char_pre=1 breaks dot repeat

### ISSUE THREAD

**Issue by kealthou** [Shougo/neocomplete.vim#332]
## Environment Information
 * OS: fedora 20
 * Vim version: 7.4 with patches 1-207, 209-475

## Minimal vimrc less than 50 lines
```
set nocompatible
set fo+=aw
set runtimepath+=~/.vim/bundle/neocomplete.vim/
let g:neocomplete#enable_at_startup = 1
let g:neocomplete#enable_auto_select = 0
set tw=10 " just for testing purposes

let g:neocomplete#enable_insert_char_pre = 1
```

## Reproducable ways from Vim starting

```
aaa222<CR>
aa11<CR>
aa11<ESC>
/11<CR>
cw22<ESC>
n.
```

## Actual results:

aa222
aa22
aa2

## Expected results:

aa222
aa22
aa22

**Comment by eed3si9n** [comment]
Thanks for the report. I was about to remove the plugin altogether.

**Comment by Shougo** [comment]
Thanks.  I will check it.
But it seems Vim's problem.  I cannot fix it.

**Comment by Shougo** [comment]
I confirmed.

```
let g:neocomplete#enable_fuzzy_completion = 0
``` Fixes it.

**Comment by oo-** [comment]
I don't think this is an issue with enable_insert_char_pre. I think fuzzy completion is broken everywhere.

Using same .vimrc as above, with enable_insert_char_pre = 0:

## Input
```
iabcde<Esc>
oabde<Esc>
.
```
## Expected
```
abcde
abde
abde
```

## Output
```
abcde
abde
ab
```

The problem disappears with enable_fuzzy_completion = 0, but it's a shame to lose that great feature.

**Comment by Shougo** [comment]
Unfortunatelly, it is Vim's dot register broken problem.
You can check it by following .vimrc.

```
set nocompatible

call append(1, 'aa11')
call append(1, 'aa11')

inoremap <F5> <C-R>=ListMonths()<CR>

func! ListMonths()
  call complete(1, ['January', 'February', 'March',
        \ 'April', 'May', 'June', 'July', 'August', 'September',
        \ 'October', 'November', 'December'])
  return ''
endfunc
```

* `/11<CR>`: Jump to "aa11"

* `cw`

*  `<F5><C-p>2<F5><C-p>2<ESC>`

*  `n.`

**Comment by oo-** [comment]
That's unfortunate. So if I understand you correctly then it's not the case that users have to choose between reliable dot-repeats and fuzzy_match, but rather that any completion plugin using vim's pums might interfere with its dot register, and fuzzy_match is so good at finding lots of buffer matches that the problem becomes more noticeable when it's on?

I guess an example of the error without fuzzy_match is something like this:
## First this input:
```
iinput/more<CR>
word<Esc>
```
## Then this input:
```
ggdwdwcwinput/more<Esc>
0j.
```
## The expected output (works with vim -u NONE, which of course never shows pum) is first:
```
input/more
word
```
## Then:
```
input/more
input/more
```
## But with neocomplete loaded it is first:
```
input/more
word
```
## Then:
```
input/more
inre
```
Do you know which specific action(s) it is that breaks the dot register? Maybe I can work around it and/or disable parts of neocomplete.

**Comment by Shougo** [comment]
> Do you know which specific action(s) it is that breaks the dot register? Maybe I can work around it and/or disable parts of neocomplete.

It is `complete()` bug.  You cannot disable it.

**Comment by nfnty** [comment]
Has this been reported upstream?

**Comment by Shougo** [comment]
No.  I want to create the patch.  But it is not easy...

**Comment by trassl** [comment]
I also have a dot repeat issue, since December. For me the first bad commit is 1f7da025efe2141a8a771289b82b71b949066dea (Use complete() instead of completefunc).

I extracted two lines, that break the dot repeat for me.

```diff
diff -x .git -r neocomplete.vim/autoload/neocomplete/mappings.vim neocomplete.vim.ok/autoload/neocomplete/mappings.vim
38c38
<           \ <C-r>=neocomplete#complete#complete_mapping()<CR><C-p><Down>
---
>           \ <C-x><C-u>
41c41
<           \ <C-r>=neocomplete#complete#complete_mapping()<CR><C-p>
---
>           \ <C-x><C-u><C-r>=neocomplete#mappings#popup_post()<CR>
```

**Comment by Shougo** [comment]
Yes.  It is known issue.

**Comment by trassl** [comment]
I do not fully understand the change in the mapping. But I was hoping, that there is a workaround (other than going back to 1f7da025efe2141a8a771289b82b71b949066dea~1).

**Comment by Shougo** [comment]
It is the reason.

https://github.com/Shougo/neocomplete.vim/issues/328

**Comment by trassl** [comment]
Thanks, now I got it.

**Comment by toupeira** [comment]
Is it possible to work around this problem with [vim-repeat](https://github.com/tpope/vim-repeat)?

**Comment by Shougo** [comment]
No.

**Comment by Shougo** [comment]
It is fixed in Vim 7.4.633

http://ftp.vim.org/vim/patches/7.4/7.4.633
